### PR TITLE
class library: add File *read

### PR DESCRIPTION
**Pull request by telephon**
Instead of requiring a callback function, it is much clearer to just
read from file and return the value.

### CODE REVIEW

File: HelpSource/Classes/File.schelp [hunk 1]
```diff
@@ -40,7 +40,22 @@ File.getcwd;
...
+argument::pathName
+the path (link::Classes/String::) that points to the file
+
+argument::mode
+the file mode to use.
+
+argument::selector
+the method to be called on the file once it is open.
+
+
```
Comment (crucialfelix) [inline HelpSource/Classes/File.schelp]: What are the valid methods that could be supplied here ?  \readAllString ,  ... ?

**Comment by crucialfelix** [comment]
Also it would be good to start each sentence with a capital letter.

File: HelpSource/Classes/File.schelp [hunk 2]
```diff
@@ -40,7 +40,22 @@ File.getcwd;
...
+open the file, evaluate the function with the file as argument, and close it again. If the process fails, close the file and throw an error.
+
+method::read
+given a path, make a new temporary file, and read the contents, close the file and return them. By default, it calls link::#-readAllString::. If the process fails, close the file and throw an error.
+
+argument::pathName
+the path (link::Classes/String::) that points to the file
+
+argument::mode
+the file mode to use.
```
Comment (telephon) [inline HelpSource/Classes/File.schelp]: yes. It may be argued that we don't need this option, but I don't know.

Comment (telephon) [inline HelpSource/Classes/File.schelp]: … I see you mean, I should add this information. Yes, I'll do so asap.

File: HelpSource/Classes/File.schelp [hunk 3]
```diff
@@ -40,7 +40,22 @@ File.getcwd;
 ::
 
 method::use
-open the file, evaluate the function with the file and close it.
+open the file, evaluate the function with the file as argument, and close it again. If the process fails, close the file and throw an error.
+
+method::read
+given a path, make a new temporary file, and read the contents, close the file and return them. By default, it calls link::#-readAllString::. If the process fails, close the file and throw an error.
```
Comment (telephon) [inline HelpSource/Classes/File.schelp]: good point, I just meant the file object!

**Comment by crucialfelix** [comment]
looks great now, thanks !

File: SCClassLibrary/Common/Files/File.sc [hunk 4]
```diff
@@ -18,6 +18,11 @@ File : UnixFILE {
 		file = this.new(pathName, mode);
 		^{ function.value(file) }.protect({ file.close });
 	}
+	*read { arg pathName, mode = "r", selector = \readAllString;
```
Comment (snappizz) [inline SCClassLibrary/Common/Files/File.sc]: How about `methodName` instead of `selector`?

Comment (jamshark70) [inline SCClassLibrary/Common/Files/File.sc]: `selector` is already a convention from the various `perform` methods.

Comment (snappizz) [inline SCClassLibrary/Common/Files/File.sc]: Understood.

Comment (telephon) [inline SCClassLibrary/Common/Files/File.sc]: yes, it's a notion that comes from the smalltalk tradition. A symbol selects a method from the receiving object's vocabulary.

File: HelpSource/Classes/File.schelp [hunk 5]
```diff
@@ -40,12 +40,35 @@ File.getcwd;
 ::
 
 method::use
-open the file, evaluate the function with the file and close it.
+Open the file, evaluate the function with the file as argument, and close it again. If the process fails, close the file and throw an error.
+
+method::read
+Given a path for an existing file, read and return the contents. By default, it calls link::#-readAllString::. If the process fails, close the file and throw an error.
```
Comment (vivid-synth) [inline HelpSource/Classes/File.schelp]: Same note as above

**Comment by adcxyz** [comment]
I agree that file opening and closing is boring admin for users.
Maybe the clearest option would be to have separate methods for all 'read entire file' methods?
Assuming text files are the default case, that could be: 
File.read("path/to/file.txt");
File.readRTF("path/to/file.rtf");
File.readHTML("path/to/file.html");
File.readSignalFile.readSignal("path/to/file.sig");

Then users need not worry about read modes, cant erase files by accident etc.

For simple formatted text files, there is also the FileReader class and its subclasses.

**Comment by snappizz** [comment]
I like this. Is it ready for merge?

**Comment by telephon** [comment]
I wonder what you think about @adcxyz 's comments. It seems like a viable simplification.

**Comment by brianlheim** [comment]
I prefer @adcxyz's suggestion; it matches the API of the existing "readAllX" instance methods.

There is now a helpfile conflict with master, but should be trivial to resolve it.

**Comment by telephon** [comment]
agreed, I'll do these.

**Comment by telephon** [comment]
Looking into it, I've come up with these two simple methods:
```
*readString { arg pathName;
		var res;
		this.use(pathName, "r", { |file| res = file.readAllString });
		^res
}
*readSignal { arg pathName;
		var res;
		this.use(pathName, "r", { |file| res = file.readAllSignal });
		^res
}
```

If you want to strip away rtf, you can write:

`File.readString(path).stripRTF`

Does that look good?

**Comment by brianlheim** [comment]
Why not rtf and html?

**Comment by telephon** [comment]
The interface `readRTF` had its purpose when the help system was still in RTF, same for HTML. So I just thought it's not really so important. Also semantically `stripRTF` seemed better, says what it does. It doesn't really give you a RTF representation back.

**Comment by brianlheim** [comment]
OTOH, it costs nothing to also make convenience functions for HTML and RTF, and this provides a more complete and consistent interface. I have personally used SC to read/strip HTML. "it's not really so important" seems like an odd argument to use given the triviality of these functions.

File: SCClassLibrary/Common/Files/File.sc (hunk 4 above)
Comment (crucialfelix) [inline SCClassLibrary/Common/Files/File.sc]: Used in ObjectiveC as well:

https://developer.apple.com/library/content/documentation/General/Conceptual/DevPedia-CocoaCore/Selector.html

**Comment by telephon** [comment]
probably it's harmless to just exactly replicate the instance read interface. For a moment I was put off by the fact that instead of just reading, it strips.

**Comment by telephon** [comment]
@brianlheim @adcxyz here is a new version, let me know if you like it, then I update the help.

The interface now simply matches the one of the instance methods:

```
*readAllString { arg pathName;
		var res;
		this.use(pathName, "r", { |file| res = file.readAllString });
		^res
	}
	*readAllSignal { arg pathName;
		var res;
		this.use(pathName, "r", { |file| res = file.readAllSignal });
		^res
	}
	*readAllStringHTML { arg pathName;
		var res;
		this.use(pathName, "r", { |file| res = file.readAllStringHTML });
		^res
	}
	*readAllStringRTF { arg pathName;
		var res;
		this.use(pathName, "r", { |file| res = file.readAllStringRTF });
		^res
	}
```

**Comment by brianlheim** [comment]
@telephon lgtm, thanks!

**Comment by brianlheim** [comment]
Help needs to be updated, and then I propose we merge :) Thanks!

**Comment by telephon** [comment]
I tried to resolve it in the online editor, but somehow my browser doesn't show the GUI properly so I couldn't save the change. Maybe someone else can try?

**Comment by brianlheim** [comment]
Still need to update the help file... it's referencing a non-existent method now.

**Comment by snappizz** [comment]
i'll take care of it

**Pull request merged**